OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,506 @@
+				/*   unknown  */ reader.ReadBytes(12);
+				LinkToPrevious = reader.ReadUInt32();
+				LinkToNext = reader.ReadUInt32();
+
+				LinkFlags = reader.ReadBytes(1)[0];
+				Volume = reader.ReadUInt16();
+				var posSave = reader.Position;
+
+				reader.Seek(tableOffset + NameOffset, SeekOrigin.Begin);
+				Filename = reader.ReadASCIIZ();
+				reader.Seek(posSave, SeekOrigin.Begin);
+			}
+		}
+
+		class CabReader : IDisposable
+		{
+			readonly FileDescriptor fileDes;
+			public uint RemainingArchiveStream;
+			public uint RemainingFileStream;
+			readonly uint index;
+			readonly string commonName;
+			ushort volumeNumber;
+			Stream cabFile;
+
+			public CabReader(FileDescriptor fileDes, uint index, string commonName)
+			{
+				this.fileDes = fileDes;
+				this.index = index;
+				this.commonName = commonName;
+				volumeNumber = fileDes.Volume;
+				--volumeNumber;
+				RemainingArchiveStream = 0;
+				if ((fileDes.Flags & FileCompressed) > 0)
+					RemainingFileStream = fileDes.CompressedSize;
+				else
+					RemainingFileStream = fileDes.ExpandedSize;
+
+				cabFile = null;
+				NextFile();
+			}
+
+			public void CopyTo(Stream dest)
+			{
+				if ((fileDes.Flags & FileCompressed) != 0)
+				{
+					var inf = new Inflater(true);
+					var buffer = new byte[165535];
+					do
+					{
+						var bytesToExtract = cabFile.ReadUInt16();
+						RemainingArchiveStream -= 2u;
+						RemainingFileStream -= 2u;
+						inf.SetInput(GetBytes(bytesToExtract));
+						RemainingFileStream -= bytesToExtract;
+						while (!inf.IsNeedingInput)
+						{
+							var inflated = inf.Inflate(buffer);
+							dest.Write(buffer, 0, inflated);
+						}
+
+						inf.Reset();
+					}
+					while (RemainingFileStream > 0);
+				}
+				else
+				{
+					do
+					{
+						RemainingFileStream -= RemainingArchiveStream;
+						dest.Write(GetBytes(RemainingArchiveStream), 0, (int)RemainingArchiveStream);
+					}
+					while (RemainingFileStream > 0);
+				}
+			}
+
+			public byte[] GetBytes(uint count)
+			{
+				if (count < RemainingArchiveStream)
+				{
+					RemainingArchiveStream -= count;
+					return cabFile.ReadBytes((int)count);
+				}
+				else
+				{
+					var outArray = new byte[count];
+					var read = cabFile.Read(outArray, 0, (int)RemainingArchiveStream);
+					if (RemainingFileStream > RemainingArchiveStream)
+					{
+						NextFile();
+						RemainingArchiveStream -= (uint)cabFile.Read(outArray, read, (int)count - read);
+					}
+
+					return outArray;
+				}
+			}
+
+			public void Dispose()
+			{
+				cabFile.Dispose();
+			}
+
+			public void NextFile()
+			{
+				if (cabFile != null)
+					cabFile.Dispose();
+
+				++volumeNumber;
+				cabFile = GlobalFileSystem.Open("{0}{1}.cab".F(commonName, volumeNumber));
+				if (cabFile.ReadUInt32() != 0x28635349)
+					throw new InvalidDataException("Not an Installshield CAB package");
+				uint fileOffset;
+				if ((fileDes.Flags & FileSplit) != 0)
+				{
+					new CommonHeader(cabFile);
```

Comment:
This is pretty ugly.  You could make it a bit cleaner (and negligibly faster) by defining a static field for `CommonHeader.Size`, and then seeking past this data instead of reading it and throwing the resulting object away.